pontoon/static/js/tags/admin/table.js

```diff
@@ -0,0 +1,172 @@
+
+    constructor (props) {
+        super(props);
+        this.visible = [];
+        this.state = {checked: new Set()};
+
+        this.handleCheckboxClick = this.handleCheckboxClick.bind(this);
+        this.handleSelectAll = this.handleSelectAll.bind(this);
+        this.handleSubmit = this.handleSubmit.bind(this);
+        this.handleTableChange = this.handleTableChange.bind(this);
+        this.handleTableSortChange = this.handleTableSortChange.bind(this);
+        this.handleTableResize = this.handleTableResize.bind(this);
+
+        this.renderResource = this.renderResource.bind(this);
+        this.renderResourceCheckbox = this.renderResourceCheckbox.bind(this);
+        this.renderSelectAllCheckbox = this.renderSelectAllCheckbox.bind(this);
+    }
+
+    get columns () {
+        // column descriptors, as expected by ReactTable
+        return [
+            {Header: this.renderSelectAllCheckbox,
+             Cell: this.renderResourceCheckbox,
+             sortable: false,
+             width: 45
+            },
+            {Header: "Resource",
+             id: "type",
+             Cell: this.renderResource
+            }];
+    }
+
+    get defaultPageSize () {
+        return 5;
+    }
+
+    get pruned() {
+        // Returns a copy of the checked set with any resource paths that are
+        // not in `this.visible` removed
+        let {checked} = this.state;
+        return new Set([...checked].filter(v => (this.visible.indexOf(v) !== -1)))
+    }
+
+    get selected () {
+        // get the pruned list of selected resources, returns all=true/false
+        // if all visible resources are selected
+        // some rows can be empty strings if there are more visible rows than
+        // resources
+        const checked = this.pruned
+        const all = (
+            checked.size > 0
+                && checked.size === this.visible.filter(v => v !== '').length);
+        return {all, checked};
+    }
+
+    get submitMessage () {
+        const {type} = this.props;
+        return type === 'assoc' ? 'Unlink resources': 'Link resources';
+    }
+
+    componentWillUpdate (prevProps) {
+        if (prevProps.data !== this.props.data) {
+            this.visible.length = 0;
+            this.setState({checked: new Set()});
+        }
+    }
+
+    handleCheckboxClick (evt) {
+        // adds/removes paths for submission
+        let {checked} = this.state;
+        checked = new Set(checked);
+        evt.target.checked ? checked.add(evt.target.name) : checked.delete(evt.target.name);
+        this.setState({checked});
+    }
+
+    handleSelectAll () {
+        // user clicked the select all checkbox...
+        // if there are some resources checked already, all are removed
+        // otherwise all visible are checked.
+        let {checked} = this.state;
+        checked = checked.size > 0 ? new Set() : new Set([...this.visible])
+        return this.setState({checked})
```

Comment:
No need to return here, right?

Reply:
i think there is - at least for testing purposes - setState is async so without this you cant check that state has changed

Reply:
I don't get it. Using `return` makes the function wait of asynchronous calls? Also, `setState` doesn't return anything.

Reply:
> return makes the function wait of asynchronous calls? 

yep, i think here it does - and certainly if you want to wait on it in tests

> setState doesn't return anything.

it returns a promise i think (except in componentdidmount iirc)

Reply:
yep im pretty confident that a) setState is async and b) you cant test against it if you dont handle the promises.

There is quite a lot of stuff on the interwebs about the async nature of setState - this seems to sum it fairly well 
 https://stackoverflow.com/questions/42038590/when-to-use-react-setstate-callback 

this also has some interesting discussion:

https://medium.com/front-end-hacking/async-await-with-react-lifecycle-methods-802e7760d802

im pretty sure that it works differently when called in `componentDidMount` but struggling to find hard evidence of that